README
NAME
    `Socket::GetAddrInfo' - address-family independent name resolving
    functions

SYNOPSIS
     use Socket qw( SOCK_STREAM );
     use Socket::GetAddrInfo qw( getaddrinfo getnameinfo );
     use IO::Socket;

     my %hints = ( socktype => SOCK_STREAM );
     my ( $err, @res ) = getaddrinfo( "www.google.com", "www", \%hints );

     die "Cannot resolve name - $err" if $err;

     my $sock;

     foreach my $ai ( @res ) {
        my $candidate = IO::Socket->new();

        $candidate->socket( $ai->{family}, $ai->{socktype}, $ai->{protocol} )
           or next;

        $candidate->connect( $ai->{addr} )
           or next;

        $sock = $candidate;
        last;
     }

     if( $sock ) {
        my ( $err, $host, $service ) = getnameinfo( $sock->peername );
        print "Connected to $host:$service\n" if !$err;
     }

DESCRIPTION
    The RFC 2553 functions `getaddrinfo' and `getnameinfo' provide an
    abstracted way to convert between a pair of host name/service name and
    socket addresses, or vice versa. `getaddrinfo' converts names into a set
    of arguments to pass to the `socket()' and `connect()' syscalls, and
    `getnameinfo' converts a socket address back into its host name/service
    name pair.

    These functions provide a useful interface for performing either of
    these name resolution operation, without having to deal with IPv4/IPv6
    transparency, or whether the underlying host can support IPv6 at all, or
    other such issues. However, not all platforms can support the underlying
    calls at the C layer, which means a dilema for authors wishing to write
    forward-compatible code. Either to support these functions, and cause
    the code not to work on older platforms, or stick to the older "legacy"
    resolvers such as `gethostbyname()', which means the code becomes more
    portable.

    This module attempts to solve this problem, by detecting at compiletime
    whether the underlying OS will support these functions. If it does not,
    the module will use pure-perl emulations of the functions using the
    legacy resolver functions instead. The emulations support the same
    interface as the real functions, and behave as close as is resonably
    possible to emulate using the legacy resolvers. See
    Socket::GetAddrInfo::Emul for details on the limits of this emulation.

    As of Perl version 5.14.0, Perl already supports `getaddrinfo' in core.
    On such a system, this module simply uses the functions provided by
    `Socket', and does not need to use its own compiled XS, or pure-perl
    legacy emulation.

EXPORT TAGS
    The following tags may be imported by `use Socket::GetAddrInfo qw( :tag
    )':

    AI      Imports all of the `AI_*' constants for `getaddrinfo' flags

    NI      Imports all of the `NI_*' constants for `getnameinfo' flags

    EAI     Imports all of the `EAI_*' for error values

    constants
            Imports all of the above constants

FUNCTIONS
  ( $err, @res ) = getaddrinfo( $host, $service, $hints )
    `getaddrinfo' turns human-readable text strings (containing hostnames,
    numeric addresses, service names, or port numbers) into sets of binary
    values containing socket-level representations of these addresses.

    When given both host and service, this function attempts to resolve the
    host name to a set of network addresses, and the service name into a
    protocol and port number, and then returns a list of address structures
    suitable to connect() to it.

    When given just a host name, this function attempts to resolve it to a
    set of network addresses, and then returns a list of these addresses in
    the returned structures.

    When given just a service name, this function attempts to resolve it to
    a protocol and port number, and then returns a list of address
    structures that represent it suitable to bind() to.

    When given neither name, it generates an error.

    The optional `$hints' parameter can be passed a HASH reference to
    indicate how the results are generated. It may contain any of the
    following four fields:

    flags => INT
            A bitfield containing `AI_*' constants. At least the following
            flags will be available:

            * `AI_PASSIVE'
              Indicates that this resolution is for a local `bind()' for a
              passive (i.e. listening) socket, rather than an active (i.e.
              connecting) socket.

            * `AI_CANONNAME'
              Indicates that the caller wishes the canonical hostname
              (`canonname') field of the result to be filled in.

            * `AI_NUMERICHOST'
              Indicates that the caller will pass a numeric address, rather
              than a hostname, and that `getaddrinfo' must not perform a
              resolve operation on this name. This flag will prevent a
              possibly-slow network lookup operation, and instead return an
              error, if a hostname is passed.

            Other flags may be provided by the OS.

    family => INT
            Restrict to only generating addresses in this address family

    socktype => INT
            Restrict to only generating addresses of this socket type

    protocol => INT
            Restrict to only generating addresses for this protocol

    Errors are indicated by the `$err' value returned; which will be
    non-zero in numeric context, and contain a string error message as a
    string. The value can be compared against any of the `EAI_*' constants
    to determine what the error is. Rather than explicitly checking, see
    also Socket::GetAddrInfo::Strict which provides functions that throw
    exceptions on errors.

    If no error occurs, `@res' will contain HASH references, each
    representing one address. It will contain the following five fields:

    family => INT
            The address family (e.g. AF_INET)

    socktype => INT
            The socket type (e.g. SOCK_STREAM)

    protocol => INT
            The protocol (e.g. IPPROTO_TCP)

    addr => STRING
            The address in a packed string (such as would be returned by
            pack_sockaddr_in)

    canonname => STRING
            The canonical name for the host if the `AI_CANONNAME' flag was
            provided, or `undef' otherwise. This field will only be present
            on the first returned address.

  ( $err, $host, $service ) = getnameinfo( $addr, $flags, $xflags )
    `getnameinfo' turns a binary socket address into a pair of
    human-readable strings, containing the host name, numeric address,
    service name, or port number.

    The optional `$flags' parameter is a bitfield containing `NI_*'
    constants. At least the following flags will be available:

    * `NI_NUMERICHOST'
      Requests that a human-readable string representation of the numeric
      address is returned directly, rather than performing a name resolve
      operation that may convert it into a hostname.

    * `NI_NUMERICSERV'
      Requests that the port number be returned directly as a number
      representation rather than performing a name resolve operation that
      may convert it into a service name.

    * `NI_NAMEREQD'
      If a name resolve operation fails to provide a name, then this flag
      will cause `getnameinfo' to indicate an error, rather than returning
      the numeric representation as a human-readable string.

    * `NI_DGRAM'
      Indicates that the socket address relates to a `SOCK_DGRAM' socket,
      for the services whose name differs between `TCP' and `UDP' protocols.

    Other flags may be provided by the OS.

    The optional `$xflags' parameter is a bitfield containing `NIx_*'
    constants. These are a Perl-level extension to the API, to indicate
    extra information.

    * `NIx_NOHOST'
      Indicates that the caller is not interested in the hostname of the
      result, so it does not have to be converted; `undef' will be returned
      as the hostname.

    * `NIx_NOSERV'
      Indicates that the caller is not interested in the service name of the
      result, so it does not have to be converted; `undef' will be returned
      as the service name.

    Errors are indicated by the `$err' value returned; which will be
    non-zero in numeric context, and contain a string error message as a
    string. The value can be compared against any of the `EAI_*' constants
    to determine what the error is. Rather than explicitly checking, see
    also Socket::GetAddrInfo::Strict which provides functions that throw
    exceptions on errors.

EXAMPLES
  Lookup for `connect'
    The `getaddrinfo' function converts a hostname and a service name into a
    list of structures, each containing a potential way to `connect()' to
    the named service on the named host.

     my %hints = ( socktype => SOCK_STREAM );
     my ( $err, @res ) = getaddrinfo( $hostname, $servicename, \%hints );
     die "Cannot getaddrinfo - $err" if $err;

     my $sock;

     foreach my $ai ( @res ) {
        my $candidate = IO::Socket->new();

        $candidate->socket( $ai->{family}, $ai->{socktype}, $ai->{protocol} )
           or next;

        $candidate->connect( $ai->{addr} )
           or next;

        $sock = $candidate;
        last;
     }

    Because a list of potential candidates is returned, the `while' loop
    tries each in turn until it it finds one that succeeds both the
    `socket()' and `connect()' calls.

    This function performs the work of the legacy functions `gethostbyname',
    `getservbyname', `inet_aton' and `pack_sockaddr_in'.

  Making a human-readable string out of an address
    The `getnameinfo' function converts a socket address, such as returned
    by `getsockname' or `getpeername', into a pair of human-readable strings
    representing the address and service name.

     my ( $err, $hostname, $servicename ) = getnameinfo( $socket->peername );
     die "Cannot getnameinfo - $err" if $err;

     print "The peer is connected from $hostname\n";

    Since in this example only the hostname was used, the redundant
    conversion of the port number into a service name may be omitted by
    passing the `NIx_NOSERV' flag.

     my ( $err, $hostname ) = getnameinfo( $socket->peername, 0, NIx_NOSERV );

    This function performs the work of the legacy functions
    `unpack_sockaddr_in', `inet_ntoa', `gethostbyaddr' and `getservbyport'.

  Resolving hostnames into IP addresses
    To turn a hostname into a human-readable plain IP address use
    `getaddrinfo' to turn the hostname into a list of socket structures,
    then `getnameinfo' on each one to make it a readable IP address again.

     my ( $err, @res ) = getaddrinfo( $hostname, "", { socktype => SOCK_RAW } );
     die "Cannot getaddrinfo - $err" if $err;

     while( my $ai = shift @res ) {
        my ( $err, $ipaddr ) = getnameinfo( $ai->{addr}, NI_NUMERICHOST, NIx_NOSERV );
        die "Cannot getnameinfo - $err" if $err;

        print "$ipaddr\n";
     }

    The `socktype' hint to `getaddrinfo' filters the results to only include
    one socket type and protocol. Without this most OSes return three
    combinations, for `SOCK_STREAM', `SOCK_DGRAM' and `SOCK_RAW', resulting
    in triplicate output of addresses. The `NI_NUMERICHOST' flag to
    `getnameinfo' causes it to return a string-formatted plain IP address,
    rather than reverse resolving it back into a hostname.

    This combination performs the work of the legacy functions
    `gethostbyname' and `inet_ntoa'.

BUILDING WITHOUT XS CODE
    In some environments it may be preferred not to build the XS
    implementation, leaving a choice only of the core or pure-perl emulation
    implementations.

     $ PERL_SOCKET_GETADDRINFO_NO_BUILD_XS=1 perl Build.PL 

BUGS
    *   Appears to FAIL on older Darwin machines (e.g. `osvers=8.11.1'). The
        failure mode occurs in t/02getnameinfo.t and appears to relate to an
        endian bug; expecting to receive `80' and instead receiving `20480'
        (which is a 16-bit `80' byte-swapped).

SEE ALSO
    *   http://tools.ietf.org/html/rfc2553 - Basic Socket Interface
        Extensions for IPv6

ACKNOWLEDGEMENTS
    Christian Hansen <chansen@cpan.org> - for help with some XS features and
    Win32 build fixes.

    Zefram <zefram@fysh.org> - for help with fixing some bugs in the XS
    code.

    Reini Urban <rurban@cpan.org> - for help with older perls and more Win32
    build fixes.

AUTHOR
    Paul Evans <leonerd@leonerd.org.uk>



Changes
Revision history for Socket-GetAddrInfo

0.21_003 CHANGES:
         * Make more of the AI_* and NI_* flags #ifdef-guarded, for OSes that
           don't provide them
         * Docs improvements - added some code examples

0.21_002 CHANGES:
         * Implement NIx_NOHOST and NIx_NOSERV; flags to getnameinfo to
           request only one of host or service name
         * No longer supports API import tags

0.21_001 CHANGES:
         * Move most exports into @EXPORT_OK
         * Provide exception-throwing version of API in
           Socket::GetAddrInfo::Strict
         * Wrap lots more flag constants
            + AI_IDN, AI_V4MAPPED, AI_ALL, AI_ADDRCONFIG
            + NI_IDN, NI_NOFQDN
            + EAI_SYSTEM, EAI_BADHINTS, EAI_PROTOCOL

0.21    BUGFIXES:
         * Exporter 5.57 is required - this has shipped with perl since 5.8.3,
           but now explicitly declared for 5.6.2
         * getservbyport() needs a protocol name on 5.6.2

0.20    CHANGES:
         * Added bin/getaddrinfo and bin/getnameinfo script (replaces previous
           examples/gai.pl)
         * Split Socket6api into its own module
         * Emulate AI_NUMERICSERV in PP emulation since most OSes in practice
           support it, even though RFC 2553 doesn't define it
         * Provide Exporter tags for named constants
         * Don't even attempt to build XS code if no C compiler is available
         * Try to use core's Socket::getaddrinfo if it exists; as it does on
           Socket 1.93 (perl 5.13.9).
         * Avoid Test::Exception and Test::Warn as they're non-core and pull
           in Sub::UpLevel
         * Use ExtUtils::CChecker 0.06's ability to write a config.h rather
           than using C commandline -DFOO that might not be portable
         * Moved legacy resolver emulation code into its own PP.pm for code
           neatness

        BUGFIXES:
         * Bugfix for non-C99 compilers - avoid C99-style struct initialiser
         * Bugfix for threaded perls - aTHX/pTHX in err_to_SV()
         * More MSWin32 fixes [rurban]
         * Use ppport.h

0.19    BUGFIXES:
         * Can't store &PL_sv_undef in an HV; store newSV(0) instead
         * Handle magic in getaddrinfo() args [chansen]
         * Win32 build fixes [chansen]

0.18    BUGFIXES:
         * Fix plain number passing for getaddrinfo()'s service on perl
           versions <5.10.0

0.17    BUGFIXES:
         * Fix 'protocol' hint to getaddrinfo()

0.16    CHANGES:
         * Accept undef as absent host/service name, as well as ""

        BUGFIXES:
         * Need -lsocket on some OSes; hunt it using ExtUtils::CChecker

0.15    CHANGES:
         * Use XSLoader instead of DynaLoader
         * import Exporter::import instead of @ISAing it
         * Use ExtUtils::CChecker for build checking

0.14    BUGFIXES:
         * Don't leak memory on getnameinfo() failures
         * Avoid possible sockaddr alignment errors by allocating/copying a
           new buffer
         (thanks to Zefram)

0.13    CHANGES:
         * Don't warn if only importing symbolic constants; test this
         * Added Test::Pod testing
         * Docs neatening in L<> links

0.12    CHANGES:
         * Added 'use warnings'
         * Added 'gai.pl' example

0.11    CHANGES:
         * Pass "socktype => SOCK_STREAM" hint in example
         * Use "return 1" rather than "exit(1)" in C code to avoid needing
           <unistd.h>; may improve builds on non-UNIX platforms

        BUGFIXES:
         * Declare ExtUtils::CBuilder as a configure_requires dependency, and
           fail gracefully if it still isn't installed when Build.PL runs

0.10    BUGFIXES:
         * Fixed a bug in getnameinfo() when handling IPv6 addresses, where
           a buffer overrun causes at least a failure from libc, but has the
           potential to cause a stack corruption or segfault.

0.09    CHANGES:
         * Conditionally skip the tests for missing hostnames if the legacy
           resolver finds an address anyway. This may mitigate test FAILs on
           smoke test machines with wildcard DNS servers.

0.08    CHANGES:
         * Declare as release

0.08_7  CHANGES:
         * Make sure to sv_2mortal() returned SV*s in XS code

        BUGFIXES:
         * Minor changes to test scripts that should reduce cornercase FAILs,
           or at least help debug them if they occur

0.08_6  BUGFIXES:
         * Set 'sa_len' member of sockaddr structure if platform needs it

0.08_5  BUGFIXES:
         * Better comparison of sockaddr returns - unpack and test components
           to avoid errors on platforms that use sin_len, since Socket doesn't
         * Use XSRETURN() in XS functions to ensure stack is correct height
         * Better diagnostic if $err tests fail - convert number to symbolic
           constant

0.08_4  BUGFIXES:
         * Define constants in XS code wrapped in #ifdef tests, to ensure
           only ones that the platform defines are attempted.
         * Print whether PP or XS implementation are being used in the
           t/00use.t file

0.08_3  BUGFIXES:
         * Use ExtUtils::CBuilder to compile the test .c file in the
           Configure() stage, to make it (hopefully) more portable
         * Some OSes return $err==0 but no results for invalid names - make
           sure the test script can cope.

0.08_2  BUGFIXES:
         * Declare variables in "PREINIT" XS section, not "INIT"; keeps
           non-C99 compilers happy
         * Skip tests that require internet connectivity unless the tester
           declares it possible
         * Declare test counts upfront in testing scripts
         * Renamed XS functions 'getaddrinfo' and 'getnameinfo'; no need to
           prefix with 'real_'
         * Folded Configure.pl inline into Build.PL, in case it has to make
           further decisions

0.08_1  CHANGES:
         * Complete rewrite to XS+PP emulation system. No longer uses Socket6
           on account of API and error reporting issues.

lib/Socket/GetAddrInfo/Emul.pm
#  You may distribute under the terms of either the GNU General Public License
#  or the Artistic License (the same terms as Perl itself)
#
#  (C) Paul Evans, 2007-2011 -- leonerd@leonerd.org.uk

package Socket::GetAddrInfo::Emul;

use strict;
use warnings;

our $VERSION = '0.21_003';

# Load the actual code into Socket::GetAddrInfo
package # hide from indexer
  Socket::GetAddrInfo;

use Carp;

use Socket;
use Scalar::Util qw( dualvar );

our @EXPORT_OK;

=head1 NAME

C<Socket::GetAddrInfo::Emul> - Pure Perl emulation of C<getaddrinfo> and
C<getnameinfo> using IPv4-only legacy resolvers

=head1 DESCRIPTION

C<Socket::GetAddrInfo> attempts to provide the C<getaddrinfo> and
C<getnameinfo> functions by some XS code that calls the real functions in
F<libc>. If for some reason this cannot be done; either there is no C
compiler, or F<libc> does not provide these functions, then they will be
emulated using the legacy resolvers C<gethostbyname>, etc... These emulations
are not a complete replacement of the real functions, because they only
support IPv4 (the C<AF_INET> socket family). In this case, the following
restrictions will apply.

=cut

# These numbers borrowed from GNU libc's implementation, but since
# they're only used by our emulation, it doesn't matter if the real
# platform's values differ
BEGIN {
   my %constants = (
       AI_PASSIVE     => 1,
       AI_CANONNAME   => 2,
       AI_NUMERICHOST => 4,
       AI_V4MAPPED    => 8,
       AI_ALL         => 16,
       AI_ADDRCONFIG  => 32,
       # RFC 2553 doesn't define this but Linux does - lets be nice and
       # provide it since we can
       AI_NUMERICSERV => 1024,

       EAI_BADFLAGS   => -1,
       EAI_NONAME     => -2,
       EAI_NODATA     => -5,
       EAI_FAMILY     => -6,
       EAI_SERVICE    => -8,

       NI_NUMERICHOST => 1,
       NI_NUMERICSERV => 2,
       NI_NOFQDN      => 4,
       NI_NAMEREQD    => 8,
       NI_DGRAM       => 16,

       # These are not gni() constants; they're extensions for the perl API /*
       NIx_NOHOST     => 1,
       NIx_NOSERV     => 2,

       # Constants we don't support. Export them, but croak if anyone tries to
       # use them
       AI_IDN                      => 64,
       AI_CANONIDN                 => 128,
       AI_IDN_ALLOW_UNASSIGNED     => 256,
       AI_IDN_USE_STD3_ASCII_RULES => 512,
       NI_IDN                      => 32,
       NI_IDN_ALLOW_UNASSIGNED     => 64,
       NI_IDN_USE_STD3_ASCII_RULES => 128,

       # Error constants we'll never return, so it doesn't matter what value
       # these have, nor that we don't provide strings for them
       EAI_SYSTEM   => -11,
       EAI_BADHINTS => -1000,
       EAI_PROTOCOL => -1001
   );

   require constant;
   constant->import( $_ => $constants{$_} ) for keys %constants;
   push @EXPORT_OK, $_ for keys %constants;
}

push @EXPORT_OK, qw(
   getaddrinfo
   getnameinfo
);

my %errstr = (
   # These strings from RFC 2553
   EAI_BADFLAGS()   => "invalid value for ai_flags",
   EAI_NONAME()     => "nodename nor servname provided, or not known",
   EAI_NODATA()     => "no address associated with nodename",
   EAI_FAMILY()     => "ai_family not supported",
   EAI_SERVICE()    => "servname not supported for ai_socktype",
);

# Borrowed from Regexp::Common::net
my $REGEXP_IPv4_DECIMAL = qr/25[0-5]|2[0-4][0-9]|1?[0-9][0-9]{1,2}/;
my $REGEXP_IPv4_DOTTEDQUAD = qr/$REGEXP_IPv4_DECIMAL\.$REGEXP_IPv4_DECIMAL\.$REGEXP_IPv4_DECIMAL\.$REGEXP_IPv4_DECIMAL/;

sub _makeerr
{
   my ( $errno ) = @_;
   my $errstr = $errno == 0 ? "" : ( $errstr{$errno} || $errno );
   return dualvar( $errno, $errstr );
}

=head2 getaddrinfo

=over 4

=item *

If the C<family> hint is supplied, it must be C<AF_INET>. Any other value will
result in an error thrown by C<croak>.

=item *

The only supported C<flags> hint values are C<AI_PASSIVE>, C<AI_CANONNAME>,
C<AI_NUMERICSERV> and C<AI_NUMERICHOST>.

The flags C<AI_V4MAPPED> and C<AI_ALL> are recognised but ignored, as they do
not apply to C<AF_INET> lookups.  Since this function only returns C<AF_INET>
addresses, it does not need to probe the system for configured addresses in
other families, so the C<AI_ADDRCONFIG> flag is also ignored.

Note that C<AI_NUMERICSERV> is an extension not defined by RFC 2553, but is
provided by most OSes. It is possible (though unlikely) that even the native
XS implementation does not recognise this constant.

=back

=cut

sub getaddrinfo
{
   my ( $node, $service, $hints ) = @_;
   
   $node = "" unless defined $node;

   $service = "" unless defined $service;

   my ( $family, $socktype, $protocol, $flags ) = @$hints{qw( family socktype protocol flags )};

   $family ||= AF_INET; # 0 == AF_UNSPEC, which we want too
   $family == AF_INET or return _makeerr( EAI_FAMILY );

   $socktype ||= 0;

   $protocol ||= 0;

   $flags ||= 0;

   my $flag_passive     = $flags & AI_PASSIVE;     $flags &= ~AI_PASSIVE;
   my $flag_canonname   = $flags & AI_CANONNAME;   $flags &= ~AI_CANONNAME;
   my $flag_numerichost = $flags & AI_NUMERICHOST; $flags &= ~AI_NUMERICHOST;
   my $flag_numericserv = $flags & AI_NUMERICSERV; $flags &= ~AI_NUMERICSERV;

   # These constants don't apply to AF_INET-only lookups, so we might as well
   # just ignore them. For AI_ADDRCONFIG we just presume the host has ability
   # to talk AF_INET. If not we'd have to return no addresses at all. :)
   $flags &= ~(AI_V4MAPPED|AI_ALL|AI_ADDRCONFIG);

   $flags & (AI_IDN|AI_CANONIDN|AI_IDN_ALLOW_UNASSIGNED|AI_IDN_USE_STD3_ASCII_RULES) and
      croak "Socket::GetAddrInfo::Emul::getaddrinfo does not support IDN";

   $flags == 0 or return _makeerr( EAI_BADFLAGS );

   $node eq "" and $service eq "" and return _makeerr( EAI_NONAME );

   my $canonname;
   my @addrs;
   if( $node ne "" ) {
      return _makeerr( EAI_NONAME ) if( $flag_numerichost and $node !~ m/^$REGEXP_IPv4_DOTTEDQUAD$/ );
      ( $canonname, undef, undef, undef, @addrs ) = gethostbyname( $node );
      defined $canonname or return _makeerr( EAI_NONAME );

      undef $canonname unless $flag_canonname;
   }
   else {
      $addrs[0] = $flag_passive ? inet_aton( "0.0.0.0" )
                                : inet_aton( "127.0.0.1" );
   }

   my @ports; # Actually ARRAYrefs of [ socktype, protocol, port ]
   my $protname = "";
   if( $protocol ) {
      $protname = getprotobynumber( $protocol );
   }

   if( $service ne "" and $service !~ m/^\d+$/ ) {
      return _makeerr( EAI_NONAME ) if( $flag_numericserv );
      getservbyname( $service, $protname ) or return _makeerr( EAI_SERVICE );
   }

   foreach my $this_socktype ( SOCK_STREAM, SOCK_DGRAM, SOCK_RAW ) {
      next if $socktype and $this_socktype != $socktype;

      my $this_protname = "raw";
      $this_socktype == SOCK_STREAM and $this_protname = "tcp";
      $this_socktype == SOCK_DGRAM  and $this_protname = "udp";

      next if $protname and $this_protname ne $protname;

      my $port;
      if( $service ne "" ) {
         if( $service =~ m/^\d+$/ ) {
            $port = "$service";
         }
         else {
            ( undef, undef, $port, $this_protname ) = getservbyname( $service, $this_protname );
            next unless defined $port;
         }
      }
      else {
         $port = 0;
      }

      push @ports, [ $this_socktype, scalar getprotobyname( $this_protname ) || 0, $port ];
   }

   my @ret;
   foreach my $addr ( @addrs ) {
      foreach my $portspec ( @ports ) {
         my ( $socktype, $protocol, $port ) = @$portspec;
         push @ret, { 
            family    => $family,
            socktype  => $socktype,
            protocol  => $protocol,
            addr      => pack_sockaddr_in( $port, $addr ),
            canonname => undef,
         };
      }
   }

   # Only supply canonname for the first result
   if( defined $canonname ) {
      $ret[0]->{canonname} = $canonname;
   }

   return ( _makeerr( 0 ), @ret );
}

=head2 getnameinfo

=over 4

=item *

If the sockaddr family of C<$addr> is anything other than C<AF_INET>, an error
will be thrown with C<croak>.

=item *

The only supported C<$flags> values are C<NI_NUMERICHOST>, C<NI_NUMERICSERV>,
C<NI_NOFQDN>, C<NI_NAMEREQD> and C<NI_DGRAM>.

=back

=cut

sub getnameinfo
{
   my ( $addr, $flags, $xflags ) = @_;

   my ( $port, $inetaddr );
   eval { ( $port, $inetaddr ) = unpack_sockaddr_in( $addr ) }
      or return _makeerr( EAI_FAMILY );

   my $family = AF_INET;

   $flags ||= 0;

   my $flag_numerichost = $flags & NI_NUMERICHOST; $flags &= ~NI_NUMERICHOST;
   my $flag_numericserv = $flags & NI_NUMERICSERV; $flags &= ~NI_NUMERICSERV;
   my $flag_nofqdn      = $flags & NI_NOFQDN;      $flags &= ~NI_NOFQDN;
   my $flag_namereqd    = $flags & NI_NAMEREQD;    $flags &= ~NI_NAMEREQD;
   my $flag_dgram       = $flags & NI_DGRAM;       $flags &= ~NI_DGRAM;

   $flags & (NI_IDN|NI_IDN_ALLOW_UNASSIGNED|NI_IDN_USE_STD3_ASCII_RULES) and
      croak "Socket::GetAddrInfo::Emul::getnameinfo does not support IDN";

   $flags == 0 or return _makeerr( EAI_BADFLAGS );

   $xflags ||= 0;

   my $node;
   if( $xflags & NIx_NOHOST ) {
      $node = undef;
   }
   elsif( $flag_numerichost ) {
      $node = inet_ntoa( $inetaddr );
   }
   else {
      $node = gethostbyaddr( $inetaddr, $family );
      if( !defined $node ) {
         return _makeerr( EAI_NONAME ) if $flag_namereqd;
         $node = inet_ntoa( $inetaddr );
      }
      elsif( $flag_nofqdn ) {
         my ( $shortname ) = split m/\./, $node;
         my ( $fqdn ) = gethostbyname $shortname;
         $node = $shortname if defined $fqdn and $fqdn eq $node;
      }
   }

   my $service;
   if( $xflags & NIx_NOSERV ) {
      $service = undef;
   }
   elsif( $flag_numericserv ) {
      $service = "$port";
   }
   else {
      my $protname = $flag_dgram ? "udp" : "tcp";
      $service = getservbyport( $port, $protname );
      if( !defined $service ) {
         $service = "$port";
      }
   }

   return ( _makeerr( 0 ), $node, $service );
}

=head1 IDN SUPPORT

This pure-perl emulation provides the IDN constants such as C<AI_IDN> and
C<NI_IDN>, but the C<getaddrinfo> and C<getnameinfo> functions will croak if
passed these flags. This should allow a program to probe for their support,
and fall back to some other behaviour instead.

=head1 AUTHOR

Paul Evans <leonerd@leonerd.org.uk>

=cut

0x55AA;
